File: 백준/Gold/20040. 사이클 게임/사이클 게임.cc
// Authored by : chjh2129
#include <bits/stdc++.h>
/*
  라운드마다 주어지는 두 점의 번호를 merge하여 같은 집합에 속하게 합니다.

  이미 같은 집합 속한 두 점에 선분을 긋는다면 사이클이 생겼다고 판단할 수 있습니다.
*/
using namespace std;

int n, m;
vector<int> p(500'005, -1);

int find(int x){ return (p[x] < 0 ? x : p[x] = find(p[x])); }

bool merge(int x, int y){
  x = find(x);
  y = find(y);
  if(x == y) return 0;
  if(p[x] == p[y]) p[x]--;
  if(p[x] > p[y]) swap(x, y);
  p[y] = x;
  return 1;
}

int main(){
  cin.tie(nullptr)->sync_with_stdio(false);

  // input
  cin >> n >> m;

  // solve
  int ans = 0;
  for(int round = 1; round <= m; round++){
    int u, v;
    cin >> u >> v;
    if(!merge(u, v)){
      ans = round;
      break;
    }
  }

  // output
  cout << ans;
}

File: 백준/Gold/20040. 사이클 게임/사이클 게임.py
#Authored by : chjh2129
import sys

input = sys.stdin.readline

def find(x: int, P: list[int]) -> int:
  if P[x] < 0: return x
  P[x] = find(P[x], P)
  return P[x]

def union(x: int, y: int, P: list[int]) -> bool:
  x = find(x, P)
  y = find(y, P)
  if x == y: return False
  
  if P[x] == P[y]: P[x] -= 1
  if P[x] > P[y]: x, y = y, x
  P[y] = x
  return True

def main() -> None:
  # input
  N, M = map(int, input().split())
  
  # Parent info
  P:list[int] = [ -1 for _ in range(N) ]

  # solve
  ans:int = 0
  for round in range(1, M+1):
    u, v = map(int, input().split())
    if not union(u, v, P):
      ans = round
      break
  
  # output
  print(ans)

sys.exit(main())